source/lac/sparse_direct.cc

```diff
@@ -928,66 +928,175 @@ SparseDirectMUMPS::initialize_matrix(const Matrix &matrix)
+        {
+          // number of nonzero elements in matrix
+          if constexpr (std::is_same_v<Matrix, SparseMatrix<double>>)
+            nnz = matrix.n_actually_nonzero_elements();
+
+          // representation of the matrix
+          a = std::make_unique<double[]>(nnz);
 
-      // number of nonzero elements in matrix
-      nnz = matrix.n_actually_nonzero_elements();
+          // matrix indices pointing to the row and column dimensions
+          // respectively of the matrix representation above (a): ie. a[k] is
+          // the matrix element (irn[k], jcn[k])
+          irn = std::make_unique<MUMPS_INT[]>(nnz);
+          jcn = std::make_unique<MUMPS_INT[]>(nnz);
 
-      // representation of the matrix
-      a = std::make_unique<double[]>(nnz);
+          size_type n_non_zero_elements = 0;
 
-      // matrix indices pointing to the row and column dimensions
-      // respectively of the matrix representation above (a): ie. a[k] is
-      // the matrix element (irn[k], jcn[k])
-      irn = std::make_unique<MUMPS_INT[]>(nnz);
-      jcn = std::make_unique<MUMPS_INT[]>(nnz);
+          // loop over the elements of the matrix row by row, as suggested in
+          // the documentation of the sparse matrix iterator class
+          if (additional_data.symmetric == true)
+            {
+              for (size_type row = 0; row < matrix.m(); ++row)
+                {
+                  for (typename Matrix::const_iterator ptr = matrix.begin(row);
+                       ptr != matrix.end(row);
+                       ++ptr)
+                    if (std::abs(ptr->value()) > 0.0 && ptr->column() >= row)
+                      {
+                        a[n_non_zero_elements]   = ptr->value();
+                        irn[n_non_zero_elements] = row + 1;
+                        jcn[n_non_zero_elements] = ptr->column() + 1;
+
+                        ++n_non_zero_elements;
+                      }
+                }
+            }
+          else
+            {
+              for (size_type row = 0; row < matrix.m(); ++row)
+                {
+                  for (typename Matrix::const_iterator ptr = matrix.begin(row);
+                       ptr != matrix.end(row);
+                       ++ptr)
+                    if (std::abs(ptr->value()) > 0.0)
+                      {
+                        a[n_non_zero_elements]   = ptr->value();
+                        irn[n_non_zero_elements] = row + 1;
+                        jcn[n_non_zero_elements] = ptr->column() + 1;
+                        ++n_non_zero_elements;
+                      }
+                }
+            }
+          id.n   = n;
+          id.nnz = n_non_zero_elements;
+          id.irn = irn.get();
+          id.jcn = jcn.get();
+          id.a   = a.get();
+        }
+    }
+  else if constexpr (std::is_same_v<TrilinosWrappers::SparseMatrix, Matrix> ||
+                     std::is_same_v<PETScWrappers::MPI::SparseMatrix, Matrix>)
```

Comment:
Is there a way to check whether the current object's MPI communicator equals the communicator used by these matrices? I suspect that all sorts of problems would arise if they are not the same.